edx/commerce-coordinator
# WIP: temp: Testing ability to configure signals and enforce signal boundaries

**Description:** Pieces of this commit can be split out into places that make sense at a later time, or we can accept the whole thing and replace all of the FIXMEs later on. This exists as a proof-of-concept that we will be able to move forward with the designs outlined in our ADRs.


**JIRA:** REV-2410

**Testing instructions:**

To test locally you can use the Django runserver command.

1. Make sure you have a virtual environment set up for Commerce Coordinator using the instructions in the repo README, and have installed all dependencies into is using `make dev_requirements` with the virtual environment active.  _([bjh] I had to `make requirements` instead to get this to work, not sure why)_
2. Start the Django test server: `python manage.py runserver localhost:8000 --settings=commerce_coordinator.settings.local`
3. You should see output ending with "Quit the server with CONTROL-C."
4. In your browser, hit the URL: http://localhost:8000/lms/test/
5. You should see JSON output indicating that two receivers were called, one successful, and the other with exception / traceback information.
6. In the shell where the server is running you should see log output indicating that two test receivers were called with the sender argument "Something".


### Local Testing With Celery

In order to use the Celery tasks included in this PR you will need to take some extra steps (since Commerce Coordinator isn't yet available in Devstack):

First install requirements into your Commerce Coordinator virtual environment *with this branch checked out*:
	
-  `make dev_requirements`

Expose the redis port for devstack (redis is the backend that Celery is using for persisting the message queue in this configuration):
- In your local Devstack directory edit devstack/docker-compose.yml add a ports section to the redis service line <s>~181</s>~243:
```
    ports:
      - "6379:6379"
```

Start redis in devstack from the Devstack directory:
- `make dev.up.redis`

Start celery from the commerce-coordinator dir:
- `celery -A commerce_coordinator worker -l INFO`

More test URLs you can hit in the browser or pipe through [jq](https://stedolan.github.io/jq/) to make the output more readable:
```
⫸ curl -s "http://localhost:8000/lms/test_celery_signal/" | jq '.'
{
  "<function test_celery_signal_task at 0x10e17a9d0>": ""
}
⫸ curl -s "http://localhost:8000/lms/test_celery_signal/" | jq '.'
{
  "<function test_celery_signal_task at 0x10e17a9d0>": ""
}
⫸ curl -s "http://localhost:8000/lms/demo_purchase_complete/" | jq '.'
{
  "<function demo_purchase_complete_order_history at 0x10e18a430>": "",
  "<function demo_purchase_complete_send_confirmation_email at 0x10e18a5e0>": "",
  "<function demo_purchase_complete_enroll_in_course at 0x10e18a670>": ""
}
```

**Author concerns:** This is very preliminary work proving out our ability to confirm and control Django signal / receiver mappings using the settings file. It's not a fully robust implementation and is just a guidepost showing that our intended implementation can work. 

Next steps I can see might be a decorator for receivers that creates a more unified logging / error handling experience, more common functions for signals, and implementing / enforcing dispatch_uid as noted here: https://docs.djangoproject.com/en/3.2/topics/signals/#preventing-duplicate-signals

We should consider whether we really want to allow signals to be sent in the Celery or management command environments; it may be easier to reason about if signals are only confined to the primary IDA and other environments could call API endpoints to trigger workflows.

We should also consider adding a management command to start the Celery worker so that it can take advantage of auto-reloading tasks when they change. Currently you have to manually stop and restart the Celery worker to pick up any code changes. This may help: https://stackoverflow.com/questions/21666229/celery-auto-reload-on-any-changes
